### PR TITLE
Modify rule S6384: Change text to the progressive education format (APPSEC-424)

### PR DESCRIPTION
**Pull request by gaetan-ferry-sonarsource**

### CODE REVIEW

File: rules/S6384/common/fix/destination.adoc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+==== Checking the intent destination
+
+Before forwarding the intent, the application can check its destination.
```
Comment (egon-okerman-sonarsource) [inline rules/S6384/common/fix/destination.adoc:3]:
```suggestion
Before forwarding the intent, the application should check its destination.
```
We can be stricter here, I think

File: rules/S6384/common/fix/destination.adoc [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+Verifying the destination package and class names belong to a list
+of accepted components can prevent most unintended use of the forwarding
+feature.
```
Comment (egon-okerman-sonarsource) [inline rules/S6384/common/fix/destination.adoc:6]:
```suggestion
Most unintended usage of the forwarding feature can be prevented by
verifying whether the destination package and class names belong to a list
of accepted components.
```

Perhaps we can also explicitly mention `Intent.resolveActivity(PackageManager pm)`.

**Review by egon-okerman-sonarsource — CHANGES_REQUESTED** [review]
I think the text could use some changes before it is merged.

File: rules/S6384/java/how-to-fix-it/android.adoc [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        ComponentName name = forward.resolveActivity(getPackageManager());
+        if (name.getPackageName().equals("package") &&
+                name.getClassName().equals("nonsensitiveclass")) {
+            startActivity(forward);
+        }
+    }
+}
+----
+
+=== How does this work?
```
[the remark this answers is not in the document]
Comment (gaetan-ferry-sonarsource) [reply rules/S6384/java/how-to-fix-it/android.adoc:40]: Yep. Those are not supported in the current config. I preferred not including them there and wait for the sonar-sec configuration to be improved.

[the remark this answers is not in the document]
Comment (egon-okerman-sonarsource) [reply rules/S6384/java/how-to-fix-it/android.adoc:40]: Fair enough

File: rules/S6384/impact.adoc [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+an external application.
+
+Therefore, an attacker exploiting an intent redirection vulnerability could
+manage to access private application's components. Depending on the features
+privately exposed, this can lead to further exploitations, sensitive data
+disclosure, or even persistent code execution.
+
+==== Information disclosure
+
+An attacker can access any non-publicly exposed component of the application.
```
[the remark this answers is not in the document]
Comment (gaetan-ferry-sonarsource) [reply rules/S6384/impact.adoc:15]: Hm I must say prefer my sentence here. There is quite a few repetitions in your version.
Can we agree on:
"An attacker can access any component of the vulnerable application, even if they are not exported" ?

File: rules/S6384/impact.adoc [hunk 5]
```diff
@@ -0,0 +1,54 @@
+=== What is the potential impact?
```
[the remark this answers is not in the document]
Comment (gaetan-ferry-sonarsource) [reply rules/S6384/impact.adoc:1]: Nice catch, I'll add this!

File: rules/S6384/impact.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (egon-okerman-sonarsource) [reply rules/S6384/impact.adoc:15]: I agree my example isn't perfect here, it was more to point out that the "how?" is not immediately clear at the moment: how can an attacker access these private components?

For example in persistent code execution, the text follows more of a structure: some components rely on dynamic code -> intent redirection can deploy malicious code -> therefore persistent code execution is reached.

File: rules/S6384/common/fix/destination.adoc (hunk 2 above)
Comment (gaetan-ferry-sonarsource) [reply rules/S6384/common/fix/destination.adoc:6]: Ok for the rewording. The resolveActivity call is already mentioned in the android.adoc file.

File: rules/S6384/impact.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (gaetan-ferry-sonarsource) [reply rules/S6384/impact.adoc:15]: I get what you mean, even if I think the two situations are different. Here, I really thought this first sentence would be a fact and introduce the actual consequence on the next paragraph. I get it's not that obvious tho.

Would this one be better for you?
"An attacker can use the affected feature as a gateway to access other components of the vulnerable application, even if they are not exported"

[the remark this answers is not in the document]
Comment (egon-okerman-sonarsource) [reply rules/S6384/impact.adoc:15]: I think that is much better!

File: rules/S6384/impact.adoc [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+The vulnerable application context is used when a vulnerable application starts
+a new component from a malicious externally provided intent. Notably, the
+restrictions applied on accessible components are lower than when accessed from
+an external application.
```
[the remark this answers is not in the document]
Comment (gaetan-ferry-sonarsource) [reply rules/S6384/impact.adoc:6]: > An affected component that forwards a malicious externally provided intent does so using the vulnerable application's context. In particular, the new component is created with the same permissions as the application and without limitations on what feature can be reached.

[the remark this answers is not in the document]
Comment (egon-okerman-sonarsource) [reply rules/S6384/impact.adoc:6]: LGTM 👍

File: rules/S6384/impact.adoc [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+==== Attack surface increase
+
+Because the attacker can access most components of the application, they can
+identify and exploit other vulnerabilities that would be present in them. The
+actual impact depends on the nested vulnerability. Exploitation probability
+depends on the in-depth security level of the application.
+
+==== Privilege escalation
+
+IF the vulnerable application has privileges on the underlying devices, an
```
Comment (egon-okerman-sonarsource) [inline rules/S6384/impact.adoc:30]:
```suggestion
If the vulnerable application has privileges on the underlying devices, an
```
Small typo

File: rules/S6384/impact.adoc [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+Therefore, an attacker exploiting an intent redirection vulnerability could
+manage to access a private application's components. Depending on the features
+privately exposed, this can lead to further exploitations, sensitive data
+disclosure, or even persistent code execution.
+
+==== Information disclosure
+
+An attacker can use the affected feature as a gateway to access other components
+of the vulnerable application, even if they are not exported. This includes
+features that handle sensitive information, even if they are not exported.
```
Comment (egon-okerman-sonarsource) [inline rules/S6384/impact.adoc:14]:
```suggestion
features that handle sensitive information.
```
Avoids repetition on the previous line

**Review by egon-okerman-sonarsource — APPROVED** [review]
There remains one typo and one repetition that should be fixed. I'll leave this up to you to finish.
Great work, the text is very extensive!

**Pull request merged**